### ISSUE TITLE
Expo dev client build failing on new architecture of RN + Hermes on Android project

### ISSUE THREAD

**Issue by mechaadi** [expo/expo#18912]
### Summary

We are trying to build the dev client on android with the new architecture of React native + Hermes and the latest versions of `react-native-gesture-handler`, `react-native-screens`, and `react-native-reanimated` but it's failing (logs below).

My guess is the file  `/node_modules/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/launcher/DevLauncherClientHost.kt` (which gets created after the ./gradlew build command) having two import statements are causing the build to break.

```
import devmenu.com.th3rdwave.safeareacontext.SafeAreaContextPackage
import devmenu.com.swmansion.gesturehandler.react.RNGestureHandlerPackage
```

It basically can't find the `devmenu` keyword. I looked for other devmenu references in the whole project directory but couldn't find anything except  maybe this `expo.modules.devmenu.DevMenuActivity` in the [Gradle file]( https://github.com/expo/expo/blob/main/packages/expo-dev-launcher/android/build.gradle#L147) of expo-dev-launcher 

### **Behavior in debug mode build, Output of ./gradlew build (gradle scan file added in stacktrace section)**

```
> Task :expo-dev-menu:compileReleaseWithDevMenuKotlin FAILED
e: /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/expo-dev-menu/android/src/debug/java/expo/modules/devmenu/DevMenuHost.kt: (16, 30): Unresolved reference: gesturehandler
e: /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/expo-dev-menu/android/src/debug/java/expo/modules/devmenu/DevMenuHost.kt: (18, 20): Unresolved reference: th3rdwave
e: /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/expo-dev-menu/android/src/debug/java/expo/modules/devmenu/DevMenuHost.kt: (35, 7): Unresolved reference: RNGestureHandlerPackage
e: /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/expo-dev-menu/android/src/debug/java/expo/modules/devmenu/DevMenuHost.kt: (37, 7): Unresolved reference: SafeAreaContextPackage
e: /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/expo-dev-menu/android/src/main/java/expo/modules/devmenu/DevMenuPackage.kt: (32, 121): Unresolved reference: DevMenuPackageDelegate
e: /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/expo-dev-menu/android/src/main/java/expo/modules/devmenu/DevMenuPackage.kt: (34, 101): Unresolved reference: DevMenuPackageDelegate
w: Detected multiple Kotlin daemon sessions at build/kotlin/sessions

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':expo-dev-menu:compileReleaseWithDevMenuKotlin'.
> Compilation error. See log for more details
```

### **Behavior in release mode build, output of ./gradlew assembleRelease (gradle scan file added in stacktrace section)**
The  `./gradlew assembleRelease` command throws another error because dev client is not required in release mode of course (logs below)

```
> Task :react-native-reanimated:configureCMakeRelWithDebInfo[arm64-v8a] FAILED
C/C++: release|arm64-v8a :CMake Error at CMakeLists.txt:74 (add_library):
C/C++: release|arm64-v8a :  Cannot find source file:
C/C++: release|arm64-v8a :    /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/Common/cpp/Fabric/FabricUtils.cpp
C/C++: release|arm64-v8a :  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
C/C++: release|arm64-v8a :  .hpp .hxx .in .txx
C/C++: release|arm64-v8a :CMake Error at CMakeLists.txt:74 (add_library):
C/C++: release|arm64-v8a :  No SOURCES given to target: reanimated
C/C++: release|arm64-v8a :CMake Generate step failed.  Build files cannot be regenerated correctly.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':react-native-reanimated:configureCMakeRelWithDebInfo[arm64-v8a]'.
> [CXX1405] error when building with cmake using /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/android/CMakeLists.txt: Build command failed.
  Error while executing process /Users/mechaadi/Library/Android/sdk/cmake/3.18.1/bin/cmake with arguments {-H/Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/android -DCMAKE_SYSTEM_NAME=Android -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DCMAKE_SYSTEM_VERSION=21 -DANDROID_PLATFORM=android-21 -DANDROID_ABI=arm64-v8a -DCMAKE_ANDROID_ARCH_ABI=arm64-v8a -DANDROID_NDK=/Users/mechaadi/Library/Android/sdk/ndk/21.4.7075529 -DCMAKE_ANDROID_NDK=/Users/mechaadi/Library/Android/sdk/ndk/21.4.7075529 -DCMAKE_TOOLCHAIN_FILE=/Users/mechaadi/Library/Android/sdk/ndk/21.4.7075529/build/cmake/android.toolchain.cmake -DCMAKE_MAKE_PROGRAM=/Users/mechaadi/Library/Android/sdk/cmake/3.18.1/bin/ninja -DCMAKE_CXX_FLAGS=-fstack-protector-all -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/android/build/intermediates/cxx/RelWithDebInfo/2l155l5p/obj/arm64-v8a -DCMAKE_RUNTIME_OUTPUT_DIRECTORY=/Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/android/build/intermediates/cxx/RelWithDebInfo/2l155l5p/obj/arm64-v8a -DCMAKE_BUILD_TYPE=RelWithDebInfo -B/Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/android/.cxx/RelWithDebInfo/2l155l5p/arm64-v8a -GNinja -DANDROID_STL=c++_shared -DNATIVE_DEBUG=false -DREACT_NATIVE_TARGET_VERSION=69 -DANDROID_TOOLCHAIN=clang -DBOOST_VERSION=1_76_0 -DNODE_MODULES_DIR=/Users/mechaadi/Documents/projects/expo/test-expo/node_modules -DJS_RUNTIME=jsc -DJS_RUNTIME_DIR=/Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native/ReactCommon/jsi -DCLIENT_SIDE_BUILD=true -DIS_NEW_ARCHITECTURE_ENABLED=true -DPLAYGROUND_APP_NAME=test-expo -DRNVERSION=69}
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /Users/mechaadi/Library/Android/sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /Users/mechaadi/Library/Android/sdk/ndk/21.4.7075529/toolchains/llvm/prebuilt/darwin-x86_64/bin/clang++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring done
  
  CMake Error at CMakeLists.txt:74 (add_library):
    Cannot find source file:
  
      /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/Common/cpp/Fabric/FabricUtils.cpp
  
    Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
    .hpp .hxx .in .txx
  
  
  CMake Error at CMakeLists.txt:74 (add_library):
    No SOURCES given to target: reanimated
  
  
  CMake Generate step failed.  Build files cannot be regenerated correctly.
```

### Managed or bare workflow?

bare

### What platform(s) does this occur on?

Android

### Package versions

"expo": "^46.0.9"
"expo-dev-client": "~1.2.1"
"expo-updates": "~0.14.5"
"react": "18.0.0"
"react-native": "0.69.5"
"react-native-gesture-handler": "^2.6.0"
"react-native-screens": "^3.17.0"
"react-native-reanimated": "^2.10.0"

### Environment

```
npx: installed 2 in 6.772s

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 12.4
      Shell: 5.8.1 - /bin/zsh
    Binaries:
      Node: 14.17.3 - /usr/local/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 6.14.13 - /usr/local/bin/npm
      Watchman: 2022.08.22.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.5, macOS 12.3, tvOS 15.4, watchOS 8.5
      Android SDK:
        API Levels: 28, 30, 31, 32, 33
        Build Tools: 29.0.2, 30.0.2, 30.0.3, 31.0.0, 32.0.0, 32.1.0, 33.0.0
        System Images: android-32 | Google APIs ARM 64 v8a, android-32 | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2021.2 AI-212.5712.43.2112.8815526
      Xcode: 13.4.1/13F100 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~46.0.9 => 46.0.9 
      react: 18.0.0 => 18.0.0 
      react-dom: 18.0.0 => 18.0.0 
      react-native: 0.69.5 => 0.69.5 
      react-native-web: ~0.18.7 => 0.18.8 
    npmGlobalPackages:
      eas-cli: 1.1.1
      expo-cli: 6.0.5
    Expo Workflow: bare
```

### Reproducible demo

https://github.com/mechaadi/expo-rn-new-arch

### Stacktrace (if a crash is involved)

Gradle scan of `gradlew build`: https://scans.gradle.com/s/qptdklze22xkm
Gradle scan of `gradlew assembleRelease`: https://scans.gradle.com/s/atjg3xzda7czk

**Comment by ajsmth** [comment]
First off I'd like to say thank you for the thoroughness of this issue. 

I was able to reproduce the build error, however the issue seems to be pointing to `react-native-reanimated`, your screenshots also seem to indicate this as well

Running `yarn expo install react-native-reanimated` downgrades to `2.9.1` - is there a reason why you are using a more recent version of reanimated? Downgrading might help resolve this build issue. 

I also noticed in your `gradle.properties`  `expo.jsEngine=jsc` which to me indicates that this build will be using JSC rather than Hermes - just thought I would point that out.

**Comment by mechaadi** [comment]
Hi @ajsmth,

Please note that the reanimated task fails in the `./gradlew assembleRelease` process because `expo-dev-client` is not needed in release builds. I have added both debug and release logs in the main issue. The `./gradlew build` process fails because of `expo-dev-launcher` which creates a file at `/node_modules/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/launcher/DevLauncherClientHost.kt` this location inside `src/debug`.

I have tried with `react-native-reanimated: 2.9.1` as well, but when I looked at the logs `./gradlew assembleRelease` I thought this is some bug with the `react-native-reanimated` library so I upgraded it to the latest version just to check if they have released any fixes for this bug, but no luck.

I tried both JSC and Hermes, build is failing in both cases on M1 Mac and windows 11.

I have added another branch with jsEngine as Hermes: https://github.com/mechaadi/expo-rn-new-arch/tree/hermes+newArch

./gradlew build logs (with Hermes enabled): https://scans.gradle.com/s/wrqdolmhojaf4 


### **EAS BUILD ISSUE (MIGHT BE UNRELATED TO THIS ISSUE)**
Also wanted to make a point that EAS builds fine with JSC in release builds (without the expo-dev-client) but `eas build fails when Hermes is enabled` with the following logs (EAS is not able to find ICU libraries, bug is reproducible using the same repo added in main issue of this thread):

```
-- Performing Test CXX_SUPPORTS_DATA_SECTIONS - Success
-- The following ICU libraries were not found:
--   uc (required)
--   i18n (required)
--   data (required)
--   uc (required)
-- Failed to find all ICU components (missing: ICU_INCLUDE_DIR ICU_LIBRARY _ICU_REQUIRED_LIBS_FOUND) (Required is at least version "52")
-- Configuring incomplete, errors occurred!
See also "/home/expo/workingdir/build/node_modules/react-native/ReactAndroid/hermes-engine/build/hermes/CMakeFiles/CMakeOutput.log".
See also "/home/expo/workingdir/build/node_modules/react-native/ReactAndroid/hermes-engine/build/hermes/CMakeFiles/CMakeError.log".
> Task :ReactAndroid:hermes-engine:configureBuildForHermes FAILED
w: Detected multiple Kotlin daemon sessions at build/kotlin/sessions
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings
508 actionable tasks: 508 executed
Unable to list file systems to check whether they can be watched. The whole state of the virtual file system has been discarded. Reason: Could not query file systems: could not open mount file (errno 2: No such file or directory)
[stderr] CMake Error at CMakeLists.txt:520 (message):
[stderr]   Unable to find ICU.
[stderr] FAILURE: Build failed with an exception.
[stderr] * What went wrong:
[stderr] Execution failed for task ':ReactAndroid:hermes-engine:configureBuildForHermes'.
[stderr] > Process 'command '/home/expo/Android/Sdk/cmake/3.18.1/bin/cmake'' finished with non-zero exit value 1
[stderr] * Try:
[stderr] > Run with --stacktrace option to get the stack trace.
[stderr] > Run with --info or --debug option to get more log output.
[stderr] > Run with --scan to get full insights.
[stderr] * Get more help at https://help.gradle.org
[stderr] BUILD FAILED in 5m 31s
Error: Gradle build failed with unknown error. See logs for the "Run gradlew" phase for more information.
```

This made me try using `JSC instead of Hermes when building with EAS`, should I create a new issue for the `eas build issue with Hermes`?

**Comment by mechaadi** [comment]
Turns out the 2.10.0 version of `react-native-reanimated` doesn't have FabricUtils shipped with it on the NPM registry. This throws the below error (basically the Common/cpp/Fabric folder is not present but the CMakeLists.txt file inside react-natitve-reanimated/android/ folder requires it):


```
 CMake Error at CMakeLists.txt:74 (add_library):
    Cannot find source file:
  
      /Users/mechaadi/Documents/projects/expo/test-expo/node_modules/react-native-reanimated/Common/cpp/Fabric/FabricUtils.cpp
  
    Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
    .hpp .hxx .in .txx
  
  
  CMake Error at CMakeLists.txt:74 (add_library):
    No SOURCES given to target: reanimated
```

This error can be solved using the `react-native-reanimated@3.0.0-rc.2` version which comes with the required Fabric files for the new architecture but this is not a complete solution because `react-native-gesture-handler` will still not work as expected (note below) and you might also get the below error.

```
    in RNGestureHandlerRootView (created by GestureHandlerRootView)
    in GestureHandlerRootView
    in Unknown (created by ExpoRoot)
    in ExpoRoot
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent), js engine: hermes
 ERROR  TypeError: Cannot read property 'stateNode' of undefined
 ```

Version 2.10.0 of `react-native-reanimated` throws a different error because of the /react-native-reanimated/android/CMakeLists.txt file 

Note: `react-native-gesture-handler` still won't work as expected in android since Fabric is still not compatible with the `RNGestureHandlerRootView`. The GestureHandlerRootView should wrap the whole app and should be called on the entry point of the application as mentioned in the [official docs](https://docs.swmansion.com/react-native-gesture-handler/docs/installation#js). Doing this doesn't throw any errors but it just simply doesn't work (you can use any gestures, the gestures will not be registered and you will see a red flash on app launch or app restarts) so I went ahead and tried to debug it with Flipper's layout inspector (img below), and I saw this `ReactUnimplementedView` node in the tree which kind of convinced me that gesture handler has some missing dependencies so I went ahead and inspected all the outer nodes in the tree and then I saw this `AppCompatTextView` which said `"'RNGestureHandlerRootView' is not Fabric compatible yet."` (img below)

![image](https://user-images.githubusercontent.com/18265351/189473758-605e0165-f0d9-4896-a2fd-9d39b16fefde.png)

Now, to solve this issue, I edited my MainActivity.java file and changed line 72 

from this:
```
reactRootView.setIsFabric(BuildConfig.IS_NEW_ARCHITECTURE_ENABLED);
```

to this:
```
reactRootView.setIsFabric(false);
```

The above solution works for both debug and release builds (new arch and hermes) of iOS and Android with the following node packages

```
"react-native-safe-area-context": "^4.3.3",
"react-native-reanimated": "2.9.1" or "2.10.0",
"react-native-gesture-handler": "^2.6.0"
"react-native-screens": "^3.17.0"
```

If, in case, the latest version of react-native-reanimated (with fabric support) is used then this solution might not work

**Comment by brentvatne** [comment]
i'm going to close this because we don't currently support fabric in expo-dev-client, but we plan to in the future, likely in sdk 48
